frontend/src/components/navigation/data-set-selector.tsx

```diff
@@ -7,35 +7,49 @@ import {
   DropdownMenuTrigger
 } from "@/components/ui/dropdown-menu.tsx";
 import { useEffect } from "react";
-import { ApiClient } from "@/lib/api.ts";
-import { authenticationProviderInstance } from "@/lib/authentication-provider.ts";
 import { useApplicationContext } from "@/lib/use-application-context.ts";
 import { SidebarMenuButton } from "@/components/ui/sidebar.tsx";
 import logoUrl from "@/assets/logo.png";
 import { Separator } from "@/components/ui/separator.tsx";
-import { useNavigate } from "react-router-dom";
-
-const api = new ApiClient(authenticationProviderInstance);
+import { useNavigate, useLocation } from "react-router-dom";
 
 export function DataSetSelector() {
-  const { dataSets, setDataSets, dataSetId, setDataSetId, account } = useApplicationContext()!;
+  const { dataSets, dataSetId, setDataSetId, account } = useApplicationContext()!;
   const navigate = useNavigate();
+  const location = useLocation();
 
   const activeDataSet = () => {
+    if (dataSetId === null && dataSets.length > 0) {
+      setDataSetId(dataSets[0].id!);
+      return dataSets[0];
+    }
     return dataSets.find((e) => e.id === dataSetId);
   };
 
   useEffect(() => {
-    const fetchData = async () => {
-      const apiDataSets = await api.dataSets().getDataSets();
-      setDataSets(apiDataSets);
-      if (!apiDataSets.find((e) => e.id === dataSetId)) {
-        setDataSetId(apiDataSets[0]?.id || null);
+    if (dataSetId === null && dataSets.length > 0) {
+      setDataSetId(dataSets[0].id!);
+    }
+  }, [dataSetId, dataSets]);
+
+  useEffect(() => {
+    const pathParts = location.pathname.split('/');
+    const dataSetIndex = pathParts.indexOf('data-sets');
+    const hasDataSetId = dataSetIndex !== -1 && !isNaN(Number(pathParts[dataSetIndex + 1]));
+
+    if (hasDataSetId) {
+      const currentDataSetId = Number(pathParts[dataSetIndex + 1]);
+      if (currentDataSetId !== dataSetId) {
+        setDataSetId(currentDataSetId);
       }
-    };
+    }
+  }, [dataSetId, location.pathname, setDataSetId]);
```

Comment:
Don't you do the same in the ApplicationContextProvider? Maybe that one is enough?